File: interpreter-for-cs/src/Runtime/Sentence/SentenceObjectDefine.cs
/*
	PeakScript: https://github.com/Jenocn/PeakScript
	By Jenocn: https://jenocn.github.io
*/

using System.Collections.Generic;

namespace peak.interpreter {

	class SentenceObjectDefine : Sentence {
		private string _name = "";
		private List<Sentence> _sentenceList = null;
		public SentenceObjectDefine(string name, List<Sentence> sentenceList) {
			_name = name;
			_sentenceList = sentenceList;
		}
		public override ExecuteResult Execute(Space space) {
			var tempSpace = new Space(SpaceType.Object);
			tempSpace.AddSpaceOfUsing(space);
			var valueObject = new ValueObject(tempSpace);
			foreach (var sentence in _sentenceList) {
				if (!IsSuccess(sentence.Execute(tempSpace))) {
					ErrorLogger.LogRuntimeError(_name);
					ErrorLogger.LogRuntimeError(ErrorRuntimeCode.ObjectDefine, "The object build failed!");
					return ExecuteResult.Failed;
				}
			}

			if (!space.AddVariable(new Variable(_name, VariableAttribute.None, valueObject))) {
				ErrorLogger.LogRuntimeError(_name);
				ErrorLogger.LogRuntimeError(ErrorRuntimeCode.ObjectDefine, "The object \"" + _name + "\" is exist!");
				return ExecuteResult.Failed;
			}

			return ExecuteResult.Successed;
		}

	}

} // namespace peak.interpreter

File: interpreter-for-cpp/src/Runtime/Sentence/SentenceObjectDefine.cpp
#include "SentenceObjectDefine.h"
#include "../Value/ValueTool.h"
#include "../Variable.h"

using namespace peak::interpreter;

SentenceObjectDefine::SentenceObjectDefine(const std::string& name, std::list<std::shared_ptr<Sentence>>& sentenceList)
	: _name(name), _sentenceList(sentenceList) {
}

ExecuteResult SentenceObjectDefine::Execute(std::shared_ptr<Space> space) {

	auto tempSpace = std::shared_ptr<Space>(new Space(SpaceType::Object));
	tempSpace->AddSpaceOfUsing(space);
	auto valueObject = std::shared_ptr<ValueObject>(new ValueObject(tempSpace));
	for (auto sentence : _sentenceList) {
		if (!IsSuccess(sentence->Execute(tempSpace))) {
			ErrorLogger::LogRuntimeError(_name);
			ErrorLogger::LogRuntimeError(ErrorRuntimeCode::ObjectDefine, "The object build failed!");
			return ExecuteResult::Failed;
		}
	}

	if (!space->AddVariable(std::shared_ptr<Variable>(new Variable(_name, VariableAttribute::None, valueObject)))) {
		ErrorLogger::LogRuntimeError(_name);
		ErrorLogger::LogRuntimeError(ErrorRuntimeCode::ObjectDefine, "The object \"" + _name + "\" is exist!");
		return ExecuteResult::Failed;
	}

	return ExecuteResult::Successed;
}
